services/docker/docker-size.service.js

```diff
@@ -0,0 +1,65 @@
+'use strict'
+
+const Joi = require('@hapi/joi')
+const prettyBytes = require('pretty-bytes')
+const { nonNegativeInteger } = require('../validators')
+const { buildDockerUrl, getDockerHubUser } = require('./docker-helpers')
+const { BaseJsonService } = require('..')
+
+const buildSchema = Joi.object({
+  name: Joi.string().required(),
+  full_size: nonNegativeInteger.required(),
+}).required()
+
+module.exports = class DockerSize extends BaseJsonService {
+  static get category() {
+    return 'size'
+  }
+
+  static get route() {
+    return buildDockerUrl('image-size', true)
+  }
+
+  static get examples() {
+    return [
+      {
+        title: 'Docker Image Size',
+        pattern: ':user/:repo',
+        namedParams: { user: 'fedora', repo: 'apache' },
+        staticPreview: this.render({ size: 126000000 }),
+      },
+      {
+        title: 'Docker Image Size (tag)',
+        pattern: ':user/:repo/:tag',
+        namedParams: { user: 'fedora', repo: 'apache', tag: 'latest' },
+        staticPreview: this.render({ size: 103000000 }),
+      },
+    ]
+  }
+
+  static get defaultBadgeData() {
+    return {
+      label: 'image size',
+      color: 'blue',
+    }
+  }
+
+  static render({ size }) {
+    return { message: prettyBytes(size) }
+  }
+
+  async fetch({ user, repo, tag }) {
+    return this._requestJson({
+      schema: buildSchema,
+      url: `https://registry.hub.docker.com/v2/repositories/${getDockerHubUser(
+        user
+      )}/${repo}/tags/${tag}`,
+      errorMessages: { 404: 'image or tag not found' },
+    })
+  }
+
+  async handle({ user, repo, tag = 'latest' }) {
```

Comment:
While doing a final review I noticed that this service is always defaulting the tag to `latest` too, and I've got the same concerns here as the `version` service. We don't necessarily to support the same fuzzy matching with the size badges that exists, but we should at least support selecting the size of the tag selected based on the same most recent semver/date criteria like the version service.

I'd be okay with addressing that in a follow up PR to able to go ahead and merge this, though I'd want to get this updated before this goes live on prod Shields.io. Any thoughts/concerns/questions with that?

Reply:
I think any sort of matching might be a little confusing in the case of the size badge. This is why I went the approach of if not tag is provided we assume `latest`, however, if the user wants to determine the size of an specific tag they can provide it explicitly.

Reply:
> I think any sort of matching might be a little confusing in the case of the size badge. This is why I went the approach of if no tag is provided we assume latest

To clarify, I am not saying we need to do the same loose/fuzzy matching that exists in the version badge. However, this badge needs to be able to provide the current size for a user-specified explicit tag as well as the size for the current tag/version that's determined dynamically without an explicit tag.

As it stands right now, this badge will only work with a user-provided explicit tag, or for users that that always publish a `latest` tag.

This is inconsistent with our other badges, and as discussed with the version badge, Shields has to support all use cases. We cannot presume/require all users are leveraging the `latest` tag pattern, as there are plenty of users that do not. 

Specifically, for the route with no user-specified/explicit tag
`../docker/image-size/user/my-repo`
where `my-repo` has these tags:
```json
[
  { "name": "1.0.0", "full_size": 2802957, "last_updated": "2020-01-18" },
  { "name": "1.2.1", "full_size": 2700000, "last_updated": "2020-02-09" },
  { "name": "3.0.0", "full_size": 3402957, "last_updated": "2020-02-02" }
]
```

In the default case we should use the size from the `3.0.0` tag, and return a badge like
![](https://img.shields.io/badge/image%20size-3.4%20MB-blue). 

That'd be impossible if the service were to always use an explicit `latest` tag like it is right now, because it would always fail with a `not found` error.

Reply:
If that's the case I'll ensure the sort param is in the size badge too, where with the aforementioned data set `1.2.1` would be selected if no tag is provided and it's date sorted, and if semver sorted then `3.0.0` is selected.

Reply:
Perfect!

Reply:
Alright, this should do it!